ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Forecast.hs

```diff
@@ -43,13 +46,31 @@ trivialForecast x = constantForecastOf () (getTipSlot x)
 -- This is primarily useful for tests; the forecast range is infinite, but we
 -- do still check the precondition, to catch any bugs.
 constantForecastOf :: a -> WithOrigin SlotNo -> Forecast a
-constantForecastOf a at = Forecast {
+constantForecastOf = constantForecastInRange Nothing
+
+-- | Forecast where the values are never changing, in a certain window.
+--
+-- This is primarily useful for tests; the forecast range is finite, and we
+-- do still check the precondition, to catch any bugs.
+constantForecastInRange :: Maybe SlotNo -> a -> WithOrigin SlotNo -> Forecast a
+constantForecastInRange range' a at = Forecast {
       forecastAt  = at
-    , forecastFor = \for ->
-                      if NotOrigin for >= at
-                        then return a
-                        else error "constantForecastOf: precondition violated"
+    , forecastFor = forecastForWithRange range'
     }
+  where
+    forecastForWithRange Nothing = \for ->
+        if NotOrigin for >= at
+          then return a
+          else error "constantForecastOf: precondition violated"
+    forecastForWithRange (Just range) = \for ->
+        let outsideForecastMaxFor = fromWithOrigin (-1) at + range + 1
```

Comment:
`SlotNo` is unsigned, so the `-1` is wrapping around to `18446744073709551615`.
```suggestion
        let outsideForecastMaxFor = succWithOrigin at + range
```

Reply:
Is it true that `succWithOrigin Origin = 0` and `succWithOrigin (At (SlotNo 0)) = 1`?

Reply:
Somehow, though, changing from `fromWithOrigin 0 at` to `fromWithOrigin (-1) at` fixed a bug. Peculiar.

Reply:
> Is it true that `succWithOrigin Origin = 0` and `succWithOrigin (At (SlotNo 0)) = 1`?

It is. Fixed in 443aed5b705a5673c9f74a9b4a70df8fb3591b4c.